File: engine/src/stub_icu.rs
pub(crate) type UChar = u16;
pub(crate) type UErrorCode = i32;

#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct UBreakIterator {
    _unused: [u8; 0],
}
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct UConverter {
    _unused: [u8; 0],
}
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct UBiDi {
    _unused: [u8; 0],
}

pub(crate) type UConverterType = i32;
pub(crate) type UBiDiLevel = u8;

pub(crate) const UCNV_UTF32_LittleEndian: UConverterType = 8;
pub(crate) const U_ZERO_ERROR: UErrorCode = 0;

pub(crate) type UBiDiDirection = u32;
pub(crate) const UBIDI_RTL: UBiDiDirection = 1;
pub(crate) const UBIDI_MIXED: UBiDiDirection = 2;

pub(crate) type UBreakIteratorType = u32;
pub(crate) const UBRK_LINE: UBreakIteratorType = 2;

extern_and_forward_stub! {
    pub(crate) fn ubidi_open => tt_ubidi_open() -> *mut UBiDi;
    pub(crate) fn ubidi_close => tt_ubidi_close(pBiDi: *mut UBiDi) -> ();
    pub(crate) fn ubidi_setPara => tt_ubidi_setPara(
        pBiDi: *mut UBiDi,
        text: *const UChar,
        length: i32,
        paraLevel: UBiDiLevel,
        embeddingLevels: *mut UBiDiLevel,
        pErrorCode: *mut UErrorCode
    ) -> ();
    pub(crate) fn ubidi_getDirection => tt_ubidi_getDirection(pBiDi: *const UBiDi) -> UBiDiDirection;
    pub(crate) fn ubidi_getVisualRun => tt_ubidi_getVisualRun(
        pBiDi: *mut UBiDi,
        runIndex: i32,
        pLogicalStart: *mut i32,
        pLength: *mut i32
    ) -> UBiDiDirection;
    pub(crate) fn ubidi_countRuns => tt_ubidi_countRuns(pBiDi: *mut UBiDi, pErrorCode: *mut UErrorCode) -> i32;
    pub(crate) fn ubrk_next => tt_ubrk_next(bi: *mut UBreakIterator) -> i32;
    pub(crate) fn ubrk_close => tt_ubrk_close(bi: *mut UBreakIterator) -> ();
    pub(crate) fn ubrk_open => tt_ubrk_open(
        type_0: UBreakIteratorType,
        locale: *const i8,
        text: *const UChar,
        textLength: i32,
        status: *mut UErrorCode
    ) -> *mut UBreakIterator;
    pub(crate) fn ubrk_setText => tt_ubrk_setText(
        bi: *mut UBreakIterator,
        text: *const UChar,
        textLength: i32,
        status: *mut UErrorCode
    ) -> ();
    pub(crate) fn ucnv_open => tt_ucnv_open(converterName: *const i8, err: *mut UErrorCode) -> *mut UConverter;
    pub(crate) fn ucnv_close => tt_ucnv_close(converter: *mut UConverter) -> ();
    pub(crate) fn ucnv_toAlgorithmic => tt_ucnv_toAlgorithmic(
        algorithmicType: UConverterType,
        cnv: *mut UConverter,
        target: *mut i8,
        targetCapacity: i32,
        source: *const i8,
        sourceLength: i32,
        pErrorCode: *mut UErrorCode
    ) -> i32;
    pub(crate) fn ucnv_fromUChars => tt_ucnv_fromUChars (
        cnv: *mut UConverter,
        dest: *mut i8,
        destCapacity: i32,
        src: *const UChar,
        srcLength: i32,
        pErrorCode: *mut UErrorCode
    ) -> i32;
    pub(crate) fn ucnv_toUChars => tt_ucnv_toUChars (
        cnv: *mut UConverter,
        dest: *mut UChar,
        destCapacity: i32,
        src: *const i8,
        srcLength: i32,
        pErrorCode: *mut UErrorCode
    ) -> i32;
}

File: engine/tectonic/stub_icu.c
/* tectonic/stub_icu.c: Binding stub for ICU functions.
   Copyright 2016-2018 the Tectonic Project
   Licensed under the MIT License.
*/

#include <unicode/ubidi.h>
#include <unicode/ubrk.h>
#include <unicode/ucnv.h>

/* first pass to make c compiler happy */
UBiDi * tt_ubidi_open(void);
void tt_ubidi_close(UBiDi *pBiDi);
void tt_ubidi_setPara(UBiDi *pBiDi, const UChar *text, int32_t length,
              UBiDiLevel paraLevel, UBiDiLevel *embeddingLevels,
              UErrorCode *pErrorCode);
UBiDiDirection tt_ubidi_getDirection(const UBiDi *pBiDi);
UBiDiDirection tt_ubidi_getVisualRun(UBiDi *pBiDi, int32_t runIndex,
                   int32_t *pLogicalStart, int32_t *pLength);
int32_t tt_ubidi_countRuns(UBiDi *pBiDi, UErrorCode *pErrorCode);
UBreakIterator* tt_ubrk_open(UBreakIteratorType type,
      const char *locale,
      const UChar *text,
      int32_t textLength,
      UErrorCode *status);
int32_t tt_ubrk_next(UBreakIterator *bi);
void tt_ubrk_close(UBreakIterator *bi);
void tt_ubrk_setText(UBreakIterator* bi,
             const UChar*    text,
             int32_t         textLength,
             UErrorCode*     status);
UConverter* tt_ucnv_open(const char *converterName, UErrorCode *err);
void tt_ucnv_close(UConverter * converter);
int32_t tt_ucnv_toAlgorithmic(UConverterType algorithmicType,
                   UConverter *cnv,
                   char *target, int32_t targetCapacity,
                   const char *source, int32_t sourceLength,
                   UErrorCode *pErrorCode);
int32_t tt_ucnv_fromUChars(UConverter *cnv,
                char *dest, int32_t destCapacity,
                const UChar *src, int32_t srcLength,
                UErrorCode *pErrorCode);
int32_t tt_ucnv_toUChars(UConverter *cnv,
              UChar *dest, int32_t destCapacity,
              const char *src, int32_t srcLength,
              UErrorCode *pErrorCode);
/* stubs */

UBiDi * tt_ubidi_open(void) {
    return ubidi_open();
}
void tt_ubidi_close(UBiDi *pBiDi) {
    return ubidi_close(pBiDi);
}
void tt_ubidi_setPara(UBiDi *pBiDi, const UChar *text, int32_t length,
              UBiDiLevel paraLevel, UBiDiLevel *embeddingLevels,
              UErrorCode *pErrorCode) {
    return ubidi_setPara(pBiDi, text, length,
              paraLevel, embeddingLevels,
              pErrorCode);
}
UBiDiDirection tt_ubidi_getDirection(const UBiDi *pBiDi) {
    return ubidi_getDirection(pBiDi);
}
UBiDiDirection tt_ubidi_getVisualRun(UBiDi *pBiDi, int32_t runIndex,
                   int32_t *pLogicalStart, int32_t *pLength) {
    return ubidi_getVisualRun(pBiDi, runIndex, pLogicalStart, pLength);
}

int32_t tt_ubidi_countRuns(UBiDi *pBiDi, UErrorCode *pErrorCode) {
    return ubidi_countRuns(pBiDi, pErrorCode);
}

UBreakIterator* tt_ubrk_open(UBreakIteratorType type,
      const char *locale,
      const UChar *text,
      int32_t textLength,
      UErrorCode *status) {
    return ubrk_open(type,
      locale,
      text,
      textLength,
      status);
}

int32_t tt_ubrk_next(UBreakIterator *bi) {
    return ubrk_next(bi);
}

void tt_ubrk_close(UBreakIterator *bi) {
    return ubrk_close(bi);
}

void tt_ubrk_setText(UBreakIterator* bi,
             const UChar*    text,
             int32_t         textLength,
             UErrorCode*     status) {
    return ubrk_setText(bi, text, textLength, status);
}

UConverter* tt_ucnv_open(const char *converterName, UErrorCode *err) {
    return ucnv_open(converterName, err);
}

void tt_ucnv_close(UConverter * converter) {
    return ucnv_close(converter);
}

int32_t tt_ucnv_toAlgorithmic(UConverterType algorithmicType,
                   UConverter *cnv,
                   char *target, int32_t targetCapacity,
                   const char *source, int32_t sourceLength,
                   UErrorCode *pErrorCode) {
    return ucnv_toAlgorithmic(algorithmicType,
                   cnv, target, targetCapacity,
                   source, sourceLength, pErrorCode);
}

int32_t tt_ucnv_fromUChars(UConverter *cnv,
                char *dest, int32_t destCapacity,
                const UChar *src, int32_t srcLength,
                UErrorCode *pErrorCode) {
	return ucnv_fromUChars(cnv,
                dest, destCapacity,
                src, srcLength,
                pErrorCode);
}

int32_t tt_ucnv_toUChars(UConverter *cnv,
              UChar *dest, int32_t destCapacity,
              const char *src, int32_t srcLength,
              UErrorCode *pErrorCode) {
	return ucnv_toUChars(cnv,
              dest, destCapacity,
              src, srcLength,
              pErrorCode);
}
